### PR TITLE
KNCOR-77: Avoid JSAPI token refresh rate limit

### PR DESCRIPTION
**Pull request by rossnelson**
On token refresh all workers were calling the oauth endpoint simultaneously.
Causing a JSAPI sql error. This update introduces a mutex lock around a new
function named refreshToken. This lock causes each worker to to wait for the
previous before refreshing its token.

Also, the refreshed token was never added to the client defaults, causing a
token refresh on every call once the token created on init expired. This update
replaces the expired token with the refreshed token on the client defaults as
well.

See [Changelog](https://github.com/knetikmedia/kne/blob/84d947983029a4feb2a7c68fd35433f8d973cacd/CHANGELOG.md) :)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,37 @@ All notable changes to this project will be documented in this file.
...
+replaces the expired token with the refreshed token on the client defaults as
+well.
+
+## Added
+- node-mutex npm package
+- kne-jsapi configs for `token_cache_key` and `token_mutex_key`
+- kne-jsapi#setupClient - If the initial setup fails, we can catch and call this
+  again passing a failure arg. Protecting KNE boot and allowing kne-jsapi to
+  reattempt a token request at a later time.
+- kne-jsapi#delCurrentToken - wrap the redis delete token funciton in a
```
Comment (jgpacheco) [inline CHANGELOG.md]: Looks like a typo `funciton` and `proimise`.

File: spec/submodules/kne-jsapi.spec.js [hunk 2]
```diff
@@ -1,18 +1,95 @@
...
+
+        it('has access to response data object', (done) => {
+            this.client.get('users/1')
+            .then(res => expect(res.result instanceof Object).toBeTruthy())
+            .then(() => done());
+        });
+
+        it('returns a error object on failure', (done) => {
+            this.client.get('users/wibble')
+            .then(res => expect(res.result instanceof Object).toBeTruthy())
```
Comment (jgpacheco) [inline spec/submodules/kne-jsapi.spec.js]: this.client.get('users/wibble')
  .then(res =>done.fail("Should fail"))
  .catch(err => {
     expect(err.constructor.name).toEqual('JsapiError'));
     done();
  });

**Comment by jgpacheco** [comment]
`"unit-test": "cross-env-shell NODE_ENV=testing CWD=\"$pwd\" jasmine",` should work across platforms.
Don't forget the dev dependency `"cross-env": "^5.1.1",`

File: bin/unit-test.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+process.argv.forEach(function (val, index, array) {
+  if (index === 2 && val.match("^spec")) { specs = [val]; }
+  if (index === 2 && !val.match("^spec")) { name = val; }
+  if (index === 3) { name = val; }
+});
+
+jasmine.loadConfigFile(process.env.CWD+'/spec/support/jasmine.json');
+jasmine.execute(specs, name);
+
```
Comment (rossnelson) [inline bin/unit-test.js]: @jgpacheco just for you my friend. :)

**Pull request merged**